pkg/utils/environment/kernel_symbols.go

```diff
@@ -2,319 +2,284 @@ package environment
 )
 
+// Kernel symbols do not have an associated size, so we define a sensible size
+// limit to prevent unrelated symbols from being returned for an address lookup
+const maxSymbolSize = 0x100000
+
+// KernelSymbol is a friendly representation of a kernel symbol.
 type KernelSymbol struct {
 	Name    string
-	Type    string
 	Address uint64
 	Owner   string
 }
-type nameAndOwner struct {
-	name  string
-	owner string
-}
-type addrAndOwner struct {
-	addr  uint64
-	owner string
-}
-
-// KernelSymbolTable manages kernel symbols with multiple maps for fast lookup.
-type KernelSymbolTable struct {
-	symbols       map[string][]*KernelSymbol
-	addrs         map[uint64][]*KernelSymbol
-	symByName     map[nameAndOwner][]*KernelSymbol
-	symByAddr     map[addrAndOwner][]*KernelSymbol
-	requiredSyms  map[string]struct{}
-	requiredAddrs map[uint64]struct{}
-	onlyRequired  bool
-	updateLock    sync.Mutex
-	updateWg      sync.WaitGroup
-}
 
-func symNotFoundErr(v interface{}) error {
-	return fmt.Errorf("symbol not found: %v", v)
+// kernelSymbolInternal is a memory efficient representation of
+// a kernel symbol, used internally for storing all symbols.
+type kernelSymbolInternal struct {
+	name string
+	// We save only the low 48 bits of the address, as all (non-percpu) symbols are at 0xffffXXXXXXXXXXXX
+	// Owner is a 16-bit index into a slice of seen owners for the symbol table this symbol belongs to.
+	// It can only be translated to the owner name if we have the symbol table.
+	// To conserve memory, we encode both of them as a single 64-bit integer where the lower 48-bits
+	// are the address and the hight 16-bits are the owner index.
+	addressAndOwner uint64
 }
 
-// NewKernelSymbolTable initializes a KernelSymbolTable with optional configuration functions.
-func NewKernelSymbolTable(opts ...KSymbTableOption) (*KernelSymbolTable, error) {
-	k := &KernelSymbolTable{}
-	for _, opt := range opts {
-		if err := opt(k); err != nil {
-			return nil, err
-		}
-	}
-
-	// Set onlyRequired to true if there are required symbols or addresses
-	k.onlyRequired = k.requiredAddrs != nil || k.requiredSyms != nil
-
-	// Initialize maps if they are nil
-	if k.requiredSyms == nil {
-		k.requiredSyms = make(map[string]struct{})
-	}
-	if k.requiredAddrs == nil {
-		k.requiredAddrs = make(map[uint64]struct{})
+func newKernelSymbolInternal(name string, address uint64, owner uint16) *kernelSymbolInternal {
+	return &kernelSymbolInternal{
+		name:            name,
+		addressAndOwner: (uint64(owner) << 48) | (address & ((1 << 48) - 1)),
 	}
+}
 
-	return k, k.Refresh()
+func (ks kernelSymbolInternal) Name() string {
+	return ks.name
 }
 
-// KSymbTableOption defines a function signature for configuration options.
-type KSymbTableOption func(k *KernelSymbolTable) error
+func (ks kernelSymbolInternal) Address() uint64 {
+	// Convert truncated address to the real kernel address
+	return (0xffff << 48) | (ks.addressAndOwner & ((1 << 48) - 1))
```

Comment:
As the prefix (all 1) overwrites owner bits I believe this can be simplified to:

```suggestion
	return kernelAddressPrefix | ks.addressAndOwner
```

Reply:
I did `return kernelAddressPrefix | (ks.addressAndOwner & addressMask)` even though as you said the prefix is all 1's, because IMO its more readable.